rs/src/main.rs

```diff
@@ -50,61 +51,199 @@ fn print(input_opts: InputOpts, output_opts: OutputOpts) -> io::Result<String> {
         return Ok(format!("{}", write::to_text(&graph, TextFormat::Full)));
     }
 
-    let graph = if output_opts.top_level {
-        let roots = graph.roots();
+    // Scope
 
-        graph.filter(|v| roots.contains(v), |_| true)
-    } else {
+    let scoped: HashSet<_> = if let Some(scope) = output_opts.scope {
         graph
+            .filter_vertices(|v| *v == scope)
+            .expand()
+            .vertices()
+            .collect()
+    } else {
+        graph.vertices().collect()
+    };
+
+    let roots: HashSet<_> = if output_opts.top_level {
+        graph.roots().vertices().collect()
+    } else {
+        graph.vertices().collect()
     };
 
+    // TODO this should now be avoidable
+    let selection: HashSet<_> = scoped.intersection(&roots).cloned().collect();
+    let graph = graph.filter_vertices(|v| selection.contains(v));
+
+    // Output
+
     let (graph, dot_format) = if !output_opts.dependencies {
         (
-            graph.filter(|_| true, |d| d == &Dependency::Strong),
+            graph.filter_edges(|d| d == &Dependency::Strong),
             DotFormat::Schedule,
         )
     } else {
         (graph, DotFormat::Dependencies)
     };
 
-    let graph = if let Some(scope) = output_opts.scope {
-        let root = vec![scope.clone()].into_iter().collect();
-        let mut tree_vertices = graph.descendants(&root);
-        tree_vertices.insert(&scope);
+    if output_opts.dot {
+        return Ok(format!("{}", write::to_dot(&graph, dot_format)));
```

Comment:
Kinda nitpicking (and not even sure if it's better) but most of these single arg formats could just use `to_string`

```suggestion
        return Ok(write::to_dot(&graph, dot_format).to_string());
```